### PR TITLE
fs: ensure ReopenWritableFile() matches rocksdb definition

### PR DESCRIPTION
**Pull request by aravind-wdc**
RocksDB expects the api ReopenWritableFile() to return an existing
file and if the file is not present, create the file and return it.
Ensure ZenFS api adheres to it.

Signed-off-by: Aravind Ramesh<Aravind.Ramesh@wdc.com>

### CODE REVIEW

**Comment by metaspace** [comment]
I am working a very similar issue for `ZenFS::DeleteFile()`. Hoping to post a patch later today.

**Comment by metaspace** [comment]
Here it is: #96

File: fs/fs_zenfs.cc [hunk 1]
```diff
@@ -516,16 +516,23 @@ IOStatus ZenFS::FileExists(const std::string& fname, const IOOptions& options,
...
   Debug(logger_, "Reopen writable file: %s \n", fname.c_str());
 
-  if (GetFile(fname) != nullptr)
+  if (zoneFile == nullptr)
     return NewWritableFile(fname, options, result, dbg);
-
-  return target()->NewWritableFile(fname, options, result, dbg);
+  else
+    result->reset(new ZonedWritableFile(zbd_, !options.use_direct_writes,
+                                        zoneFile, &metadata_writer_));
```
Comment (metaspace) [inline fs/fs_zenfs.cc:534]: We have some overlap of code in `NewWritableFile()`, `ReopenWritableFile()`. Could we not make both delegate to a protected `OpenFile()`? I think that is what the posix layer is doing, we could model after that.

**Review by metaspace — CHANGES_REQUESTED** [review]

File: fs/fs_zenfs.cc (hunk 1 above)
Comment (aravind-wdc) [reply fs/fs_zenfs.cc]: Thanks, good suggestion. Done.

**Comment by aravind-wdc** [comment]
ReuseWritableFile() is also not as per rocksdb definition, fixed it.

**Comment by yhr** [comment]
@aravind-wdc :What is wrong with the current implementation of ReuseWritableFile? Your patch will make returned writablefiles append at the old file's end.

Reading the posix implementation, ReuseWritableFile() re-opens the current file and **starts overwriting the file at offset 0**.
We can't do overwrites in zenfs, so the current implementation in zenfs deletes the old file and creates a new file with the new name. That's ok. 

ReuseWriteableFile seems to be a bit of a hack to avoid new allocations in regular file systems for performance reasons.

**Comment by aravind-wdc** [comment]
@yhr In the rocksdb file_system.h, for ReuseWritableFile(), it says, 
"Reuse an existing file by renaming it and opening it as writable."
https://github.com/facebook/rocksdb/blob/main/include/rocksdb/file_system.h#L367

ReuseWritableFile() has a new file name and an old-filename as the arguments, fname and old_fname respectively.
As per my understanding, the caller expects it to rename the existing file "old_fname" to "fname" and simply return the newly renamed file "fname".

In current ZenFS implementation of ReuseWritableFile() we check for the new file "fname" instead of the existing file "old_fname", since it is new file name, it will not be found, we return NotFound always. 
If it is found we call NewWritableFile(), which basically deletes it and creates a new file. Since we are checking for a non-existing file(fname), so NewWritableFile() is never called.

This patch, checks if old_file is present, if not, it returns that it is not found.
If found, it renames it to old_fname, and reopens the file and returns it (not delete the file)

Please let me know if I am mistaken, I will recall this patch

**Comment by aravind-wdc** [comment]
If it's ok to delete it, then we need to change the GetFile(fname) to GetFile(old_fname) and retain the same behavior of deleting the file.

**Comment by metaspace** [comment]
> If it's ok to delete it, then we need to change the GetFile(fname) to GetFile(old_fname) and retain the same behavior of deleting the file.

@aravind-wdc There is a default implementation in `FileSystem` that we inherit if we don't implement our own:
```c++
  auto closedirDeleter = [](DIR* d) {
    if (d != nullptr) closedir(d);
  };
  std::unique_ptr<DIR, decltype(closedirDeleter)> aux_dir{
      FLAGS_aux_path.c_str(), std::move(closedirDeleter)};
```

**Comment by metaspace** [comment]
> @aravind-wdc :What is wrong with the current implementation of ReuseWritableFile?

@yhr The method is supposed to delete `old_fname`. Our implementation does not do that. Its bugged:
```c++
if (GetFile(fname) == nullptr)
  return IOStatus::NotFound("Old file does not exist");

return NewWritableFile(fname, file_opts, result, dbg);
```

**Comment by metaspace** [comment]
> This patch, checks if old_file is present, if not, it returns that it is not found.
> If found, it renames it to old_fname, and reopens the file and returns it (not delete the file)

@aravind-wdc That looks good to me

**Comment by aravind-wdc** [comment]
> @aravind-wdc What is the actual required semantics? Should the file be truncated after return from `ReuseWritableFile` or should writes continue from the old write pointer?

The actual semantics are not very clear. In the rocksdb/include/file_system.h , it says that the file has to be renamed and returned. But not whether it should be truncated.
I don't think the posix implementation truncates it, i need to check again.
I thought it should continue writing from old WP, until Hans mentioned that it has to be truncated.

**Comment by aravind-wdc** [comment]
> ```c++
> IOStatus s = RenameFile(old_fname, fname, opts.io_options, dbg);
> if (!s.ok()) {
>   return s;
> }
> return NewWritableFile(fname, opts, result, dbg);
> ```

This is what I have done, along with a error log saying "rename failed"

**Comment by yhr** [comment]
> > This patch, checks if old_file is present, if not, it returns that it is not found.
> > If found, it renames it to old_fname, and reopens the file and returns it (not delete the file)
> 
> @aravind-wdc What is the actual required semantics? Should the file be truncated after return from `ReuseWritableFile` or should writes continue from the old write pointer?
> 
> > Reading the posix implementation, ReuseWritableFile() re-opens the current file and starts overwriting the file at offset 0.
> 
> @yhr where do you see this? It is not in the default implementation in `FileSystem::ReuseWritableFile` as far as I can tell, not in `PosixFileSystem::ReuseWritableFile` either.

@metaspace : Opening the file without O_APPEND will make writes to overwrite from file beginning, right? see the file open flags set in the posix ReuseWritableFile.

**Comment by yhr** [comment]
And sorry, I did not look at the right piece of code when checking what we are doing now in PosixWritableFile.

The current implementation IS broken. But I do believe that the right thing to do here is to delete the old file if it exists and open a new file for writing using the new file name. We can't do in-place overwrites (as it looks like the posix implementation does)

**Comment by aravind-wdc** [comment]
> But I do believe that the right thing to do here is to delete the old file if it exists and open a new file for writing using the new file name. We can't do in-place overwrites (as it looks like the posix implementation does)

Yes, we do not have to rename and reopen the file.
Changed the behavior to as explained above.

**Comment by yhr** [comment]
Fixed in https://github.com/westerndigitalcorporation/zenfs/pull/129

**Pull request closed without merging**